### PR TITLE
Track branches by polling

### LINKED ISSUE #50: Add support to track Git repo branches by polling / using webhooks to act on branch updates
In order to use a branch as a desired state for a Git repo, we'll want to add the ability to poll branches to receive updates through webhooks.

This approach is used to not have stale branches after an update, and for the operator to continuously drive towards the commit state of the tracked branch.

### PR DESCRIPTION
**Pull request by lblackstone**
### Proposed changes
Poll the specified git repo every 60 seconds and automatically
deploy the latest commit if it has changed.

### Related issues (optional)

Fix #50

### CODE REVIEW

**Comment by lblackstone** [comment]
@viveklak I'm still working on the reconciliation logic, but this change will deploy the latest commit from a branch. Does this general approach make sense to you?

**Comment by lblackstone** [comment]
One open question I have is whether we should also track the default (master) branch if no branch or commit is specified. I currently only implemented tracking if a branch is explicitly set.

Alternatively, we could expose a flag to toggle branch tracking, but it seems like the tracking behavior is probably what users would expect if they explicitly opt into a branch rather than a commit.

**Comment by vivekpulumi** [comment]
> One open question I have is whether we should also track the default (master) branch if no branch or commit is specified. I currently only implemented tracking if a branch is explicitly set.
> 
> Alternatively, we could expose a flag to toggle branch tracking, but it seems like the tracking behavior is probably what users would expect if they explicitly opt into a branch rather than a commit.

I think requiring a branch is reasonable. Currently the comments in the stack CR specification are our defacto documentation. Could you update the comments for branch and commit to clarify this behavior: https://github.com/pulumi/pulumi-kubernetes-operator/blob/master/pkg/apis/pulumi/v1alpha1/stack_types.go#L98?

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,13 +2,12 @@ CHANGELOG
 =========
 
 ## HEAD (Unreleased)
-(None)
+
+- Automatically track git branches without a specified commit [#162](https://github.com/pulumi/pulumi-kubernetes-operator/pull/162)
```
Comment (vivekpulumi) [inline CHANGELOG.md:6]:
```suggestion
- Automatically track git branches without a specified commit.  [#162](https://github.com/pulumi/pulumi-kubernetes-operator/pull/162). If a branch is specified but no commit, the operator will now poll the repository every minute to look for new changes.
```

**Comment by vivekpulumi** [comment]
We might also need to call out that every commit is not necessarily reflected by the operator in this mode. e.g. if multiple commits in short succession, the operator will only try to reflect the latest at the time of polling.

**Review by vivekpulumi — APPROVED** [review]

**Comment by vivekpulumi** [comment]
CC @elsesiy for comments/thoughts.

**Review by mikhailshilkov — APPROVED** [review]
I'm not super familiar with the operator but the change LGTM. I guess I'd expend `master` to work the same way by default but I'll leave this to your judgment.

**Comment by lblackstone** [comment]
> I guess I'd expend master to work the same way by default but I'll leave this to your judgment.

I opened https://github.com/pulumi/pulumi-kubernetes-operator/issues/164 to track the likely resolution to this difference.

File: pkg/controller/stack/stack_controller.go [hunk 2]
```diff
@@ -209,6 +209,20 @@ func (r *ReconcileStack) Reconcile(request reconcile.Request) (reconcile.Result,
 		}
 	}
 
+	// If a branch is specified, then track changes to the branch.
+	trackBranch := len(sess.stack.Branch) > 0
```
Comment (viveklak) [inline pkg/controller/stack/stack_controller.go:213]: Any reason why the change to disable branch tracking if an explicit commit is specified?

Comment (lblackstone) [reply pkg/controller/stack/stack_controller.go:213]: I realized that `branch` and `commit` are mutually exclusive, so we should only need to check for the presence of `branch`.

https://github.com/pulumi/pulumi-kubernetes-operator/blob/597053a08e0081dfc4a4e9f2877467d28de91bb9/pkg/apis/pulumi/v1alpha1/stack_types.go#L96-L101
